lib/auth/grpcserver.go

```diff
@@ -4369,6 +4369,96 @@ func (g *GRPCServer) AppendDiagnosticTrace(ctx context.Context, in *proto.Append
 	return connectionDiagnosticV1, nil
 }
 
+// GetKubernetesCluster returns the specified kubernetes cluster resource.
+func (g *GRPCServer) GetKubernetesCluster(ctx context.Context, req *types.ResourceRequest) (*types.KubernetesClusterV3, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	kubeCluster, err := auth.GetKubernetesCluster(ctx, req.Name)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	kubeClusterV3, ok := kubeCluster.(*types.KubernetesClusterV3)
+	if !ok {
+		return nil, trace.BadParameter("unsupported kubernetes cluster type %T", kubeCluster)
+	}
+	return kubeClusterV3, nil
+}
+
+// CreateKubernetesCluster creates a new kubernetes cluster resource.
+func (g *GRPCServer) CreateKubernetesCluster(ctx context.Context, cluster *types.KubernetesClusterV3) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cluster.SetOrigin(types.OriginDynamic)
+	if err := auth.CreateKubernetesCluster(ctx, cluster); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &empty.Empty{}, nil
+}
+
+// UpdateKubernetesCluster updates existing kubernetes cluster resource.
+func (g *GRPCServer) UpdateKubernetesCluster(ctx context.Context, cluster *types.KubernetesClusterV3) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cluster.SetOrigin(types.OriginDynamic)
+	if err := auth.UpdateKubernetesCluster(ctx, cluster); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &empty.Empty{}, nil
+}
+
+// GetKubernetesClusters returns all kubernetes cluster resources.
+func (g *GRPCServer) GetKubernetesClusters(ctx context.Context, _ *empty.Empty) (*types.KubernetesClusterV3List, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clusters, err := auth.GetKubernetesClusters(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var kubeClusters []*types.KubernetesClusterV3
```

Comment:
Nit: could preallocate `kubeClusters` since you know how many elements you'll be putting in it